# KUIS 2 Pembelajaran Mesin

## Tentang dataset

Pada sebuah lembaga kemanusiaan berbasis teknologi informasi, dibentuk beberapa tim kemanusiaan yang bertugas untuk terjun ke masyarakat memberikan bantuan dalam bentuk aplikasi komputer atau layanan pelatihan IT.  Kegiatn tersebut dilaksanakan setiap tahun dengan ketentuan anggota organisasi tersebut dapat bergabung ke beberapa tim sebagai ketua atau anggota. Terdapat pula anggota yang sudah tidak aktif di organisasi tersebut (terhapus di database mereka —> label N/A).  Organisasi tersebut juga melakukan rekrutmen volunteer, namun tidak semua tim menggunakan bantuan volunteer. Hasil kegiatan dari setiap tim dapat diwujudkan ke dalam sebuah pengakuan hak cipta atau publikasi karya ilmiah atau publikasi berita di media massa.

## Ketentuan
1. Pengerjaan secara individu alias setiap mahasiswa!
2. Terdapat data dengan nilai #N/A, pertimbangkan cara terbaik untuk mengatasi nilai tersebut.
3. Tidak boleh ada duplikasi pekerjaan antar mahasiswa
4. Berikan penjelasan untuk setiap tahapan. Jangan hanya kode!

## Soal
1. Lakukan pemilahan data training dan testing yang sesuai dan proporsional dari data tersebut! (10 poin)
2. Tentukan data validasi! (10 poin)
3. Tentukan secara tepat kolom yang menjadi "label" (20 poin)
4. Terapkan dua jenis metode klasifikasi dan lakukan perbandingan akurasi  dengan wajib melibatkan variabel keterlibatan volunteer sampai variabel luaran-luaran yang dihasilkan! (20 poin)
5. Lakukan tunning hyperparameter berdasarkan hasil klasifikasi pada nomor 3 menggunakan metode Grid Search dan Random Search! Simpulkan parameter apa yang paling baik digunakan berdasarkan metode klasifikasi yang dipilih, dan berapa akurasinya. (40 poin)

# Jawaban Quiz 2
### Dewangga Putra (2041720222)

In [1]:
import pandas as pd

data = pd.read_excel('data/dataset_quiz.xlsx')

data.head(100)

id_ketua  id_anggota1  id_anggota2  id_anggota3  id_anggota4  id_anggota5  \
0     7206.0       7214.0       6211.0       8602.0       7801.0          NaN   
1     7214.0       7109.0       8402.0       7903.0       9202.0          NaN   
2     5911.0       8201.0       8701.0       9101.0       9003.0          NaN   
3     8402.0       7702.0       8602.0       8903.0       8903.0          NaN   
4     8201.0       8305.0       5911.0       5903.0       8402.0          NaN   
..       ...          ...          ...          ...          ...          ...   
80    8803.0       8804.0       9102.0       8005.0       9303.0          NaN   
81    6211.0       8901.0       9003.0       9303.0       9007.0          NaN   
82    8601.0       8903.0       8703.0          NaN          NaN          NaN   
83    8703.0       8601.0       9202.0       9102.0       6204.0          NaN   
84    8102.0       7109.0       5903.0       7206.0       8903.0          NaN   

   keterlibatan_vltr  jml_vltr_terlibat  tahun    skema luaran_hki  \
0              TIDAK                  0   2019  REGULER  TIDAK ADA   
1              TIDAK                  0   2019  REGULER  TIDAK ADA   
2              TIDAK                  0   2019  REGULER  TIDAK ADA   
3              TIDAK                  0   2019  REGULER  TIDAK ADA   
4              TIDAK                  0   2019  REGULER  TIDAK ADA   
..               ...                ...    ...      ...        ...   
80                YA                  3   2022  REGULER  TIDAK ADA   
81                YA                  3   2022  REGULER  TIDAK ADA   
82                YA                  3   2022  REGULER  TIDAK ADA   
83                YA                  3   2022  REGULER  TIDAK ADA   
84             TIDAK                  0   2022  REGULER  TIDAK ADA   

   jenis_luaran_hki potensi_hki luaran_lain jenis_luaran_lain  \
0         TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
1         TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
2         TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
3         TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
4         TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
..              ...         ...         ...               ...   
80        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
81        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
82        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
83        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   
84        TIDAK ADA       TIDAK   TIDAK ADA         TIDAK ADA   

   status_luaran_lain  
0           TIDAK ADA  
1           TIDAK ADA  
2           TIDAK ADA  
3           TIDAK ADA  
4           TIDAK ADA  
..                ...  
80          TIDAK ADA  
81          TIDAK ADA  
82          TIDAK ADA  
83          TIDAK ADA  
84          TIDAK ADA  

[85 rows x 16 columns]

In [2]:
# melakukan backup data ke var dtest
dtest = data.copy()

In [3]:
# menghapus kolom id_anggota5 karena bernilai null
dtest = dtest.drop(dtest.columns[[5]], axis=1)

In [4]:
# Cek kolom null
dtest.isnull().sum()


id_ketua               3
id_anggota1            5
id_anggota2            4
id_anggota3            3
id_anggota4           13
keterlibatan_vltr      0
jml_vltr_terlibat      0
tahun                  0
skema                  0
luaran_hki             0
jenis_luaran_hki       0
potensi_hki            0
luaran_lain            0
jenis_luaran_lain      0
status_luaran_lain     0
dtype: int64

In [5]:
# mengganti value NA(null) ke 0
dtest =dtest.fillna(0)

In [6]:
#Cek kembali apakah masih terdapat kolom null
dtest.isnull().sum()

id_ketua              0
id_anggota1           0
id_anggota2           0
id_anggota3           0
id_anggota4           0
keterlibatan_vltr     0
jml_vltr_terlibat     0
tahun                 0
skema                 0
luaran_hki            0
jenis_luaran_hki      0
potensi_hki           0
luaran_lain           0
jenis_luaran_lain     0
status_luaran_lain    0
dtype: int64

In [7]:
# cek data duplikasi
dtest.duplicated().sum()

0

In [8]:
# Cek informasi sata
print(dtest.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85 entries, 0 to 84
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id_ketua            85 non-null     float64
 1   id_anggota1         85 non-null     float64
 2   id_anggota2         85 non-null     float64
 3   id_anggota3         85 non-null     float64
 4   id_anggota4         85 non-null     float64
 5   keterlibatan_vltr   85 non-null     object 
 6   jml_vltr_terlibat   85 non-null     int64  
 7   tahun               85 non-null     int64  
 8   skema               85 non-null     object 
 9   luaran_hki          85 non-null     object 
 10  jenis_luaran_hki    85 non-null     object 
 11  potensi_hki         85 non-null     object 
 12  luaran_lain         85 non-null     object 
 13  jenis_luaran_lain   85 non-null     object 
 14  status_luaran_lain  85 non-null     object 
dtypes: float64(5), int64(2), object(8)
memory usage: 10.1+ KB
N

In [9]:
dtb = dtest.copy() # backup data

#melakukan seleksi fitur seluruh kolom kecual kolom id
dtb = dtb.drop(dtb.iloc[:,:5], axis=1)
dtb = dtb.drop(dtb.iloc[:,2:3], axis=1)

dtb.head(100)

keterlibatan_vltr  jml_vltr_terlibat    skema luaran_hki jenis_luaran_hki  \
0              TIDAK                  0  REGULER  TIDAK ADA        TIDAK ADA   
1              TIDAK                  0  REGULER  TIDAK ADA        TIDAK ADA   
2              TIDAK                  0  REGULER  TIDAK ADA        TIDAK ADA   
3              TIDAK                  0  REGULER  TIDAK ADA        TIDAK ADA   
4              TIDAK                  0  REGULER  TIDAK ADA        TIDAK ADA   
..               ...                ...      ...        ...              ...   
80                YA                  3  REGULER  TIDAK ADA        TIDAK ADA   
81                YA                  3  REGULER  TIDAK ADA        TIDAK ADA   
82                YA                  3  REGULER  TIDAK ADA        TIDAK ADA   
83                YA                  3  REGULER  TIDAK ADA        TIDAK ADA   
84             TIDAK                  0  REGULER  TIDAK ADA        TIDAK ADA   

   potensi_hki luaran_lain jenis_luaran_lain status_luaran_lain  
0        TIDAK   TIDAK ADA         TIDAK ADA          TIDAK ADA  
1        TIDAK   TIDAK ADA         TIDAK ADA          TIDAK ADA  
2        TIDAK   TIDAK ADA         TIDAK ADA          TIDAK ADA  
3        TIDAK   TIDAK ADA         TIDAK ADA          TIDAK ADA  
4        TIDAK   TIDAK ADA         TIDAK ADA          TIDAK ADA  
..         ...         ...               ...                ...  
80       TIDAK   TIDAK ADA         TIDAK ADA          TIDAK ADA  
81       TIDAK   TIDAK ADA         TIDAK ADA          TIDAK ADA  
82       TIDAK   TIDAK ADA         TIDAK ADA          TIDAK ADA  
83       TIDAK   TIDAK ADA         TIDAK ADA          TIDAK ADA  
84       TIDAK   TIDAK ADA         TIDAK ADA          TIDAK ADA  

[85 rows x 9 columns]

In [10]:
# melakukan labelling pada kolom
from sklearn.preprocessing import LabelEncoder

labelEn = LabelEncoder()
for col in dtb.columns:
    dtb[col] = labelEn.fit_transform(dtb[col])

display(dtb.head())

keterlibatan_vltr  jml_vltr_terlibat  skema  luaran_hki  jenis_luaran_hki  \
0                  0                  0      1           1                 1   
1                  0                  0      1           1                 1   
2                  0                  0      1           1                 1   
3                  0                  0      1           1                 1   
4                  0                  0      1           1                 1   

   potensi_hki  luaran_lain  jenis_luaran_lain  status_luaran_lain  
0            2            0                  3                   1  
1            2            0                  3                   1  
2            2            0                  3                   1  
3            2            0                  3                   1  
4            2            0                  3                   1

In [11]:
#melakukan seleksi fitur
X = dtb.iloc[:, 1:].values
y = dtb['keterlibatan_vltr'].values

X.shape

(85, 8)

In [12]:
# melakukan split dan training data
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=42)

In [13]:
# metode klasifikasi dengan Decision Tree
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# menyesuaikan dt ke set training
dt = DecisionTreeClassifier()
dt.fit(X_train, y_train)

# Memprediksi label set test
y_pred_dt = dt.predict(X_test)

# Menghitung set accuracy
acc_dt = accuracy_score(y_test, y_pred_dt)
print("Test set accuracy: {:.2f}".format(acc_dt))
print(f"Test set accuracy: {acc_dt}")

Test set accuracy: 0.95
Test set accuracy: 0.9534883720930233


In [14]:
# metode klasifikasi dengan RandomForest
from sklearn.ensemble import RandomForestClassifier

# menyesuaikan rf ke set training
rf = RandomForestClassifier(n_estimators=30, random_state=42)
rf.fit(X_train, y_train)

# Memprediksi label set test
y_pred_rf = rf.predict(X_test)

# Menghitung set accuracy
acc_rf = accuracy_score(y_test, y_pred_rf)
print("Test set accuracy: {:.2f}".format(acc_rf))
print(f"Test set accuracy: {acc_rf}")

Test set accuracy: 0.95
Test set accuracy: 0.9534883720930233


In [15]:
from sklearn.model_selection import RepeatedKFold

# inisiasi repated k-fold
cv = RepeatedKFold(n_splits=4, n_repeats=3, random_state=42)

In [16]:
# Gridsearch menggunakan DecisionTree
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV

# inisiasi model
dte = DecisionTreeClassifier()

# Melakukan pedfinisian hyperparameter yang akan digunakan
params = {
    'criterion': ['gini', 'entropy', 'log_loss'],
    'max_depth': list(range(5,11)),
}

# inisiasi grid berdasarkan nilai repeated k-fold dan hyperparameter
grid = GridSearchCV(dte, param_grid=params, cv=cv, n_jobs=-1)

%timeit grid.fit(X_train, y_train)

# melakukan evaluasi dengan score
score = grid.score(X_train, y_train)

print(f'Hasil evaluasi: {score}')
print(f'Konfigurasi hyperparameter: {grid.best_params_}')

94.8 ms ± 2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Hasil evaluasi: 1.0
Konfigurasi hyperparameter: {'criterion': 'gini', 'max_depth': 5}


In [17]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import RandomizedSearchCV

dtr = DecisionTreeClassifier()

param = {
    'criterion': ['gini', 'entropy'],
    'max_depth': list(range(5,11))
}

# inisiasi randoms berdasarkan nilai repeated k-fold dan hyperparameter
randoms = RandomizedSearchCV(dtr, param_distributions=param, cv=cv, n_jobs=-1)

%timeit randoms.fit(X_train, y_train)

# Evaluasi dengan score
score = randoms.score(X_train, y_train)

print(f'Hasil evaluasi: {score}')
print(f'Konfigurasi hyperparameter: {randoms.best_params_}')

60.8 ms ± 2.72 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
Hasil evaluasi: 1.0
Konfigurasi hyperparameter: {'max_depth': 8, 'criterion': 'entropy'}


In [23]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score
 
mnb = MultinomialNB()

mnb.fit(X_train, y_train)

# melakukan prediksi data train
y_pred_train = mnb.predict(X_train)
acc_train = accuracy_score(y_train, y_pred_train)

# melakukan prediksi data test
y_pred_test = mnb.predict(X_test)
acc_test = accuracy_score(y_test, y_pred_test)

print(f'Hasil akurasi data train: {acc_train}')
print(f'Hasil akurasi data test: {acc_test}')


Hasil akurasi data train: 0.9761904761904762
Hasil akurasi data test: 0.9767441860465116


In [24]:
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

svc = SVC(kernel="linear")

svc.fit(X_train, y_train)

# melakukan prediksi data train
y_pred_trains = svc.predict(X_train)
acc_trains = accuracy_score(y_train, y_pred_trains)

# melakukan prediksi data test
y_pred_tests = svc.predict(X_test)
acc_tests = accuracy_score(y_test, y_pred_tests)

print(f'Hasil akurasi data train: {acc_trains}')
print(f'Hasil akurasi data test: {acc_tests}')

Hasil akurasi data train: 0.9761904761904762
Hasil akurasi data test: 0.9767441860465116


### Kesimpulan 

Jadi kesimpulan dari hasil tuning hyperparameter yakni untuk metode <b><i>GridSearch</i></b> mendapatkan score evaluasi <b>1.00</b> dengan kofigurasi parameter terbaik adalah <i>{'criterion': 'gini', 'max_depth': 5}</i> </br></br>
Sedangkan untuk metode <b><i>RandomSearch</i></b> mendapatkan score evaluasi sama yaitu <b>1.00</b> dengan kofigurasi parameter terbaik adalah <i>{'criterion': 'entropy', 'max_depth': 7}</i> </br></br>
Saya juga melakukan cek evaluasi skor data training dan testing dengan menggunakan 2 metode yaitu  <b><i>Multinomial</i></b> dan <b><i>SVM</i></b> dan keduanya menunjukkan hasil yang sama yaitu <b>0.9761904761904762</b> untuk akurasi data train dan <b>0.9767441860465116</b> untuk akurasi data test